### ISSUE TITLE
Support Java 17

### ISSUE THREAD

**Issue by marek-parfianowicz** [openclover/clover#183]
Java 17 is another LTS (long-term support) release of Java. 

Add support for new language features, check binary compatibility.

**Comment by heruan** [comment]
Is there any 4.6 preview to use? We just moved to JDK 17 and Clover stopped working 😢

**Comment by marek-parfianowicz** [comment]
Hi! The 4.5.0 version came with an experimental support for Java 17. The 4.5.1 contains one regression bug, caused by the 'record' keyword. 

I also implemented even more missing code constructs on the master branch. The only missing thing is handling of switch expressions. I encourage you to build 4.5.2 version from sources (the master branch has 4.5.2-SNAPSHOT). 

https://github.com/openclover/clover/issues?q=is%3Aissue+milestone%3A4.5.2

I plan to release 4.5.2 and 4.6.0 next quarter. Cheers!

**Comment by ottlinger** [comment]
Just ran into the same problem as the new instanceof notation stops clover from working, code example:
```java
        if (authentication != null && authentication.getPrincipal() instanceof ApplicationUser userDetails) {

```
yields a broken build with:
`Caused by: clover.antlr.NoViableAltException: unexpected token: userDetails`
looking very much forward to a release candidate for JDK17.

**Comment by marek-parfianowicz** [comment]
This has been fixed in issue #214. It's awaiting release 4.5.2, hopefully this month.
________________________________

**Comment by marek-parfianowicz** [comment]
4.5.2 has been released, it supports full Java 17 language set.
